pkg/build/generator/generator.go

```diff
@@ -0,0 +1,243 @@
+package generator
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+// BuildGenerator is a central place responsible for generating new Build objects
+// from BuildConfigs and other Builds.
+type BuildGenerator struct {
+	Client GeneratorClient
+}
+
+type GeneratorClient interface {
+	GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuild(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+type Client struct {
+	GetBuildConfigFunc     func(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfigFunc  func(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuildFunc           func(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuildFunc        func(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepositoryFunc func(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+func (c Client) GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error) {
+	return c.GetBuildConfigFunc(ctx, name)
+}
+
+func (c Client) UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error {
+	return c.UpdateBuildConfigFunc(ctx, buildConfig)
+}
+
+func (c Client) GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error) {
+	return c.GetBuildFunc(ctx, name)
+}
+
+func (c Client) CreateBuild(ctx kapi.Context, build *buildapi.Build) error {
+	return c.CreateBuildFunc(ctx, build)
+}
+
+func (c Client) GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error) {
+	return c.GetImageRepositoryFunc(ctx, name)
+}
+
+// Generate returns new Build object based on BuildRequest object
+func (g *BuildGenerator) Generate(ctx kapi.Context, request *buildapi.BuildRequest) (*buildapi.Build, error) {
+	var newBuild *buildapi.Build
+	if len(request.BuildConfigName) > 0 {
+		glog.Infof("Generating build for config %s", request.BuildConfigName)
+		bc, err := g.Client.GetBuildConfig(ctx, request.BuildConfigName)
+		if err != nil {
+			return nil, err
+		}
+		newBuild, err = g.generateBuildFromConfig(ctx, bc, request.Revision)
+		if err != nil {
+			return nil, err
+		}
+		err = g.Client.UpdateBuildConfig(ctx, bc)
```

Comment:
so fundamentally you still have a race condition here because 2 Generate calls can be made simultaneously, unless etcd is going to reject the 2nd update.  And if etcd is going to reject the 2nd update, this should be a retryable failure.

(also if etcd does not reject the 2nd update, then you've also got the possibility of clobbering other updates the the BC like an update to the LastImageId or an update the user was making via CLI)